README
Version 0.1

Pronounced /'zɛfiːmæp/ or ZE-fee-map.

zephymap is a utility to receive zephyr notifications of mail on IMAP servers.
This is acheived by running zephymap.py on a computer or server of your choice
and having it check the IMAP servers you indicate in your config file (see
below) and notifying you of new messages via zephyr notification on a class of
your choice.

**NOTE**: Due to the specifics of implementation, a SIGINT (Ctrl-C) will not
kill zephymap. The proper signal to send zephymap is SIGQUIT (Ctrl-\).


INSTALLING LIBRARIES
================================================================================
zephymap requires python-zephyr in order to send notifications, which requires
libzephyr-dev, which may or may not require comerr-dev. libzephyr-dev is known
to work on linerva, and is in the Ubuntu/Debian repositories.

To install python-zephyr:
1. Determine which version of python you are using via 'python --version'; 2.5
   and 2.6 are both known to work.

2. Run 

   mkdir -p ~/.local/lib/python2.5/site-packages

   (replacing ~/.local with a directory of your choice, and 2.5 with 2.6 if
   appropriate).

3. Open your .bashrc_profile or other suitable dotfile (e.g., .bashrc.mine) and
   add the line

   export PYTHONPATH=~/.local/lib/python2.5/site-packages:$PYTHONPATH

   where the directory is the one you made in step 2.

3. Either download python-zephyr from:
   
   http://github.com/ebroder/python-zephyr/downloads

   and unpack the archive, or (if you feel like using the development version)
   clone the git repo by running:
   
   git clone git://github.com/ebroder/python-zephyr.git

   Either way, cd into the created python-zephyr directory.

4. Run

   python setup.py install --prefix=~/.local

   where ~/.local is the directory containing the lib/ directory from step
   2. This should install python-zephyr.

INSTALLING ZEPHYMAP
================================================================================
zephymap itself requires no special setup.

1. Create a directory into which to download zephymap.

2. Download a copy, either from

   http://github.com/veinor/zephymap/downloads

   or get the latest development build (not guaranteed to be stable or even
   work) via

   git clone git://github.com/veinor/zephymap.git
   
3. Set up the ~/.zephymap.conf configuration file as described in the
   CONFIGURATION section

4. Run zephymap:
   
   python zephymap.py

5. Subscribe to the class zephymap is sending zephyrs to (default MAIL). In
   barnowl,

   :sub MAIL * me

   will do this, where 'me' is replaced with your username.

CONFIGURATION
================================================================================
zephymap reads from a config file in your home directory named ".zephymap.conf".
This file must be of the following format:

[zephyr]
recipient: joeuser
[Mailbox 1]
server: imap.foo.com
username: username1
password: password1
[Mailbox 2]
server: imap.baz.com
username: username2
ssl: true

The [mailbox] section names (the string inside the []) should contain a short
meaningful name for the mail server it defines. This will appear in the
instance of zephyrs sent by zephymap.

The [zephyr] section has the following (case sensitive) options and must be
named such:

recipient   The zephyr user to whom zephymap should make the zephyrs readable.

class       The class to which zephymap will send mail notification zephyrs.
            This is the class to which the user identified in "recipient" will
            need to be subscribed in order to see the zephyrs. The default is
            "MAIL".

Each IMAP [mailbox] section has the following (caseinsensitive) options:

server      The URI of the IMAP server (e.g., "imap.gmail.com" for Gmail and
            "imap.exchange.mit.edu" for MIT's Exchange email server)

username    The username for the account on the indicated server.

password    The password for the account on the indicated server. Leaving this
            option out will cause zephymap to prompt you to input your password
            at startup. This is the suggested behaviour for the security
            concious.

interval    An interval value in seconds that indicates the time to wait between
            checking the mailbox for new messages. It cannot be set less than 10
            seconds. The default is 20 seconds.

ssl         A boolean value that indicates whether or not the server utilizes
            ssl. Default is true.

port        An integer value that indicates the port to be used to access the IMAP
            server. The default for an SSL server is 993, and the default for a
            non-SSL server is 143.

include     A semi-colon separated list of folder names to be included. If left
            blank, the default is all folders. Subfolders can be included by
            indicating "folder/subfolder"; for example,

            MIT/SIPB; zephymap; INBOX

            Note that spaces in the middle of folder names are important, but
            spaces at the beginning or end are not.


exclude     A semi-colon separated list of folder names to be excluded even if
            they exist in the include list. The default is to exclude none. Syntax
            is identical to that of the "include" option.

regex       A boolean option that causes include and exclude to interpret their
            arguments as Perl-style regular expressions. Use this option only if
            you are familiar with regular expressions. Note that the regex will
            automatically be case-insensitive

Contact Patrick at zephymap@mit.edu or on (MIT) zephyr via the zephymap class.


CHANGELOG
Version 0.1
----------------------------------------------
* Added configuring sleep timers.
* More robust debugging messages.
* Allowed non-standard IMAP/SSL ports.
* Reworked the README
* Filtering checked folders, both by list and by regex.

Version 0.0.3.4
----------------------------------------------
* Correctly correctly handle the random abort bug.

Version 0.0.3.3
----------------------------------------------
* Correctly handle the random abort bug.

Version 0.0.3.2
----------------------------------------------
* Fix idiotic typo in README related to installing python-zephyr

Version 0.0.3.1
----------------------------------------------
* Added information on how to kill zephymap

Version 0.0.3
----------------------------------------------
* Fixed a bug that would randomly crash zephymap
* Threaded out each mailbox checker
* Allow user to input password at each startup

Version 0.0.2.1
----------------------------------------------
* Proper port for non-SSL

Version 0.0.2
----------------------------------------------
* SSL now default
* Supports messages in multiple folders
  (e.g., gmail tags)
* Filter messages by regex
* Fix various typos in the README

Version 0.0.1
----------------------------------------------
* Initial release!



emailparser.py
#!/usr/bin/python
import imaplib, time, email.utils, email, calendar
import re
from socket import sslerror
from sys import stderr

class EmailParser:
    def __init__(self, server, username, password, port=None, use_ssl=False, last_check=time.time(), regex=".*"):
        self.server = server
        if not port:
            if use_ssl: self.port = 993
            else: self.port = 143
        else:
            self.port = port
        if use_ssl: self.imap = imaplib.IMAP4_SSL(self.server, self.port)
        else: self.imap = imaplib.IMAP4(self.server, self.port)
        self.username = username
        self.password = password
        self.imap.login(self.username, self.password)
        self.last_check = last_check
        self.regex = regex
        self.set_last_uids()
        
        
    def set_last_uids(self):
        self.last_uid = {}
        for folder in self.get_folders():
            if "[Gmail]" in folder: continue
            selection = self.imap.select(folder, True) # open read-only
            if selection[0] == "NO": continue
            if selection[1][0] == '0':
                self.last_uid[folder] = 0
                continue
            last_msgid = selection[1][0]
            uid_text = self.imap.fetch(last_msgid, "UID")[1][0]
            self.last_uid[folder] = int(re.search("\(UID (\d+)\)", uid_text).group(1))
            
    def get_folders(self):
        # folders are indicated like (\\HasNoChildren) "." "INBOX.Foo"; we just want INBOX.Foo
        folder_re = re.compile(r'\(.*?\) ".*" (?P<name>.*)')
        predicate_re = re.compile(self.regex)
        folders = [folder_re.match(f_str).groups()[0].strip('"')
                for f_str in self.imap.list()[1]]
        return [f for f in folders if predicate_re.match(f)]
    
    def check(self):
        """
        Check for messages received since the last check.
        Return the number of unread messages.
        """
        headers = []
        try:
            for folder in self.get_folders():
                if "[Gmail]" in folder: continue
                response, [nmesgs] = self.imap.select(folder, True)  # open read-only
                if response == "NO": continue
                # XXX: large number is because * will always return the last message
                throwaway, new = self.imap.search(None, 'UNSEEN', "(UID %d:99999999)" % (self.last_uid[folder] + 1))
                if new == ['']: continue # skip all-read folders
                print "Checking folder %s at %s." % (folder, time.asctime())
                indices = ','.join(new[0].split(' '))

                # for some reason, I get )s mixed in with actual header/response pair information.
                new_headers = [email.message_from_string(x[1]) for x in self.imap.fetch(indices, "(BODY[HEADER])")[1] if x != ')']
                for new_header in new_headers: new_header["folder"] = folder
                headers += new_headers
                uid_text = self.imap.fetch(nmesgs, "UID")[1][0]
                self.last_uid[folder] =  int(re.search("\(UID (\d+)\)", uid_text).group(1))
                
        except (imaplib.IMAP4.abort, sslerror), e:
            # Okay. There's this stupid bug in the SSL library that I don't feel like finding the cause of
            # that means that sometimes I get a random EOF. I don't know what state it leaves the connection
            # in, so just reinitialize everything.
            # Also, sometimes imaplib will just randomly abort. Again, fuck it and reinitialize.
            if isinstance(e, sslerror) and e[0] != 8:
                raise

            if isinstance(self.imap, imaplib.IMAP4_SSL):
                self.imap = imaplib.IMAP4_SSL(self.server)
            else:
                self.imap = imaplib.IMAP4(self.server)
            self.imap.login(self.username, self.password)
            if isinstance(e, sslerror):
                print >> stderr, "SSL bug in server %s at %s." % (self.server, time.asctime())
            elif isinstance(e, imaplib.IMAP4.abort):
                print >> stderr, "abort bug in server %s at %s." % (self.server, time.asctime())
            return self.check()
            
            
        return headers


emailhandler.py
#!/usr/bin/python
import imaplib, email.utils, email, calendar, time
import re
from socket import sslerror
from sys import stderr
import logging

logger = logging.getLogger("emailhandler")
logger.setLevel(logging.DEBUG)

class EmailHandler:
    def __init__(self, server, username, password, port=None, use_ssl=False, include=".*", exclude="^$"):
        self.server = server
        if not port: # default ports
            if use_ssl: self.port = 993
            else: self.port = 143
        else:
            self.port = port
        if use_ssl: self.imap = imaplib.IMAP4_SSL(self.server, self.port)
        else: self.imap = imaplib.IMAP4(self.server, self.port)
        self.username = username
        self.password = password
        self.imap.login(self.username, self.password)

        self.include = re.compile(include, re.I)
        self.exclude = re.compile(exclude, re.I)
        self.set_last_uids()
        
        
    def set_last_uids(self):
        """
        Set the last-seen UIDs for each folder.
        """
        
        self.last_uid = {}
        for folder in self.get_folders():
            if "[Gmail]" in folder: continue # Avoid checking Gmail's special folders.
            selection = self.imap.select(folder, True) # open read-only
            if selection[0] == "NO": continue
            last_msgid = selection[1][0]
            if last_msgid == '0': # No messages found.
                self.last_uid[folder] = 0
                continue
            uid_text = self.imap.fetch(last_msgid, "UID")[1][0] # looks like (UID 1234)
            self.last_uid[folder] = int(re.search("\(UID (\d+)\)", uid_text).group(1))
            
    def get_folders(self):
        """
        Get the list of folders.
        """
        # folders are indicated like (\\HasNoChildren) "." "INBOX.Foo"; we just want INBOX.Foo.
        # that middle thing is a separator, which we can ignore since we don't care about nesting.
        folder_re = re.compile(r'\(.*?\) "(?P<sep>.*)" (?P<name>.*)')
        matches = [folder_re.match(f_str).groups() for f_str in self.imap.list()[1]]
        logger.debug("Found folders %s." % [match[1].strip('"') for match in matches])
        canonical_folders = [match[1].strip('"').replace(match[0], "/") for match in matches]
        matching_folders = [folder for folder in canonical_folders
                            if self.include.search(folder) and not self.exclude.search(folder)]
        logger.debug("Matched folders %s." % matching_folders)
        return matching_folders

    def check(self):
        """
        Check for messages received since the last check.
        Return the number of unread messages.
        """
        headers = []
        try:
            for folder in self.get_folders():
                if "[Gmail]" in folder: continue
                response, [nmesgs] = self.imap.select(folder, True)  # open read-only
                if response == "NO": continue
                # XXX: large number is because * will always return the last message
                throwaway, new = self.imap.search(None, 'UNSEEN', "(UID %d:99999999)" % (self.last_uid[folder] + 1))
                if new == ['']: continue # skip all-read folders
                logger.info("Checking folder %s on server %s." % (folder, self.server))

                indices = new[0].replace(' ', ',') # it gives me '2 3 4', but it requires '2,3,4'. god I hate IMAP.
                
                # for some reason, I get )s mixed in with actual header/response pair information.
                new_headers = [email.message_from_string(x[1]) for x in self.imap.fetch(indices, "(BODY[HEADER])")[1] if x != ')']
                logger.info("Found %d message%s in folder %s on server %s." % (len(new_headers), "" if len(new_headers) == 1 else "s",
                                                                               folder, self.server))
                for new_header in new_headers: new_header["folder"] = folder # tag with folder information
                headers += new_headers
                uid_text = self.imap.fetch(nmesgs, "UID")[1][0] # mark that we read the UID
                self.last_uid[folder] = int(re.search("\(UID (\d+)\)", uid_text).group(1))
                
        except (imaplib.IMAP4.abort, sslerror), e:
            # Okay. There's this stupid bug in the SSL library that I don't feel like finding the cause of
            # that means that sometimes I get a random EOF. I don't know what state it leaves the connection
            # in, so just reinitialize everything.
            # Also, sometimes imaplib will just randomly abort. Again, fuck it and reinitialize.
            if isinstance(e, sslerror) and e[0] != 8:
                raise

            if isinstance(self.imap, imaplib.IMAP4_SSL):
                self.imap = imaplib.IMAP4_SSL(self.server)
            else:
                self.imap = imaplib.IMAP4(self.server)
            self.imap.login(self.username, self.password)
            
            if isinstance(e, sslerror):
                logger.warning("SSL bug in server %s at %s." % (self.server, time.asctime()))
            elif isinstance(e, imaplib.IMAP4.abort):
                logger.warning("abort bug in server %s at %s." % (self.server, time.asctime()))

            # Try it again
            return self.check()
            
            
        return headers


zephymap.py
#!/usr/bin/python

import ConfigParser, os.path
import sys
from emailhandler import EmailHandler
import zephyr
import time
import getpass
from threading import Thread
import logging
import re
import getopt

version = "0.1"
config_file = "~/.zephymap.conf"
global_section = "zephyr"
logger = logging.getLogger("zephymap")
logger.setLevel(logging.DEBUG)
default_interval = 20
usage = """
usage: zephymap.py [-v|-V]
must have a ~/.zephymap.conf file. 
For configuration and installation instructions, see the README.
"""

class EmailThread(Thread):
    def __init__(self, handler, name, interval=20):
        Thread.__init__(self, name=name)
        self.handler = handler
        self.interval = interval
        self.logger = logging.getLogger("zephymap.%s" % name)

    def run(self):
        while True:
            self.check()
            time.sleep(self.interval)

    def check(self):
        logger = self.logger
        logger.info("Checking for messages.")
        msgs = self.handler.check() # grab message headers
        # some e-mail providers support tagging, resulting in dupe messages
        # in folders. we only want to notify once, so we group by message ID.
        msg_groups = group(msgs, lambda x: x["Message-ID"])
        n_mesgs = len(msg_groups)
        logger.info("%d message%s." % (n_mesgs,  "" if n_mesgs == 1 else "s"))
        
        for msg_id in msg_groups:
            msg_group = msg_groups[msg_id]
            folders = ','.join([msg["folder"] for msg in msg_group]) # all folders the message is in
            msg = msg_group[0] # the only difference is the folder, so 0 is as good as any
            instance_name = "%s.%s" % (self.getName(), folders) # e.g., Gmail.INBOX
            body = """
You have new mail in %s.

From: %s
Subject: %s
To: %s
""".strip() % (instance_name, msg["From"], msg["Subject"], msg["To"])
            logger.info("Sending notification to instance %s, subject %s." % (instance_name, msg["Subject"]))
            zephyr.ZNotice(cls=target_class, instance=instance_name, fields=[msg_id, body],
                           recipient=target, sender="zephymap", isPrivate=True).send()
        

def load_config():
    """
    Load handlers and other appropriate state.
    Returns a dict of handler identifier -> EmailHandler object
    """
    # globals shared among all handlers
    global target_class
    global target
    global default_interval

    scp = ConfigParser.SafeConfigParser()

    # bail on a bad read
    logger.info("Loading config.")
    if (scp.read(os.path.expanduser(config_file)) == []):
        print "Failed to read config file at %s." % config_file
        sys.exit(1)

    handlers = {}

    target = scp.get(global_section, "recipient")
    target_class = "MAIL"
    if scp.has_option(global_section, "class"):
        target_class = scp.get(global_section, "class")

    if scp.has_option(global_section, "interval"):
        default_interval = scp.getint(global_section, "interval")

    # loop through each account
    for section in scp.sections():
        if section == global_section: continue # zephyr is for globals
        
        logger.info("Parsing section %s." % section)
        kwargs = {}
        kwargs['username'] = username = scp.get(section, "username")
        kwargs['server'] = server = scp.get(section, "server")
        if scp.has_option(section, "password"):
            password = scp.get(section, "password")
        else:
            password = getpass.getpass("Password for server %s, username %s: " % (server, username))

        kwargs['password'] = password
            
        # determine whether to use ssl; defaults to yes
    
        if scp.has_option(section, "ssl") and not scp.getboolean(section, "ssl"):
            ssl = False
            default_port = 143
        else:
            ssl = True
            default_port = 993

        kwargs['use_ssl'] = ssl

        if scp.has_option(section, "port"):
            port = scp.getint("port")
        else:
            port = default_port

        kwargs['port'] = port

        if scp.has_option(section, "interval"):
            interval = scp.getint(section, "interval")
        else:
            interval = default_interval

        interval = max(interval, 10) # don't want to be constantly checking.


        if scp.has_option(section, "regex"):
            is_regex = scp.get(section, "regex")
        else:
            is_regex = False

        if scp.has_option(section, "include"):
            include = scp.get(section, "include")
            if is_regex: include_re = include
            else: include_re = clude_to_re(include)
        else: include_re = "^.*$"

        if scp.has_option(section, "exclude"):
            exclude = scp.get(section, "exclude")
            if is_regex: exclude_re = exclude
            else: exclude_re = clude_to_re(exclude)
        else: exclude_re = "^$"

        kwargs['include'] = include_re
        kwargs['exclude'] = exclude_re

        
        logger.debug("Constructing handler for section %s: server %s, username %s, ssl %s, port %d, interval %d, is_regex %s, include_re %s, exclude_re %s."
                     % (section, server, username, ssl, port, interval, is_regex, include_re, exclude_re))
        handlers[section] = EmailHandler(**kwargs)
        handlers[section].interval = interval

    return handlers    
        

def group(things, f):
    f_dict = {}
    f_vals = list(set(map(f, things))) # unique values of f[thing]
    for f_val in f_vals:
        f_dict[f_val] = filter(lambda x: f(x) == f_val, things)
    return f_dict

def clude_to_re(s):
    elements = s.split(";")
    regexified = [r'(^%s(/.*)?$)' % re.escape(elem.strip()) for elem in elements]
    return "|".join(regexified)
    

if __name__ == "__main__":
    root_logger = logging.getLogger()
    ch = logging.StreamHandler()

    ch.setLevel(logging.WARN)

    try:
        opts, args = getopt.getopt(sys.argv[1:], "vV", ["help", "version"])
    except getopt.GetoptError, e: # Unknown option passed
        print str(e)
        print usage
        sys.exit(2)
    for o, a in opts:
        if o == "-v": ch.setLevel(logging.INFO)
        elif o == "-V": ch.setLevel(logging.DEBUG)
        elif o == "--help":
            print usage
            sys.exit()
        elif o == "--version":
            print version
            sys.exit()
        else:
            assert False, "unknown, unhandled option"

    # strictly, I should pad to the longest thread name length, but that would require me to
    # know that information before I print anything out, and I want to print when I'm loading
    # a config file.
    ch.setFormatter(logging.Formatter("%(asctime)s - %(name)20s - %(levelname)8s - %(message)s"))
    root_logger.addHandler(ch)

    logger.info("Initializing zephyr.")
    zephyr.init()
    handlers = load_config()

    for name in handlers:
        t = EmailThread(handlers[name], name, handlers[name].interval)
        t.start()
